src/service/video-manager-impl.js

```diff
@@ -284,43 +299,111 @@ class VideoEntry {
- * @return {boolean}
+ * Detects whether autoplay is supported.
+ * Note that even if platfrom supports autoplay, users or browsers can disable
+ * autoplay to save data / battery. This function detects both platfrom support
+ * and when autoplay is disabled.
+ *
+ * Service dependencies are taken explicitly for testability.
+ *
+ * @private visible for testing.
+ * @param {!./ampdoc-impl.AmpDoc} ampdoc
+ * @param {!./platform-impl.Platform} platform
+ * @param {!./timer-impl.Timer} timer
+ * @param {boolean} isLiteViewer
+ * @return {!Promise<boolean>}
  */
-function platformSupportsAutoplay(win) {
-  // Do not support autoplay in amp-lite viewer
-  if (getMode(win).lite) {
-    return false;
+export function supportsAutoplay(ampdoc, platform, timer, isLiteViewer) {
+
+  // Use cached result if available.
+  if (supportsAutoplayCache_) {
+    return supportsAutoplayCache_;
   }
 
-  const platform = platformFor(win);
-  // non-mobile platforms always supported autoplay
-  if (!platform.isAndroid() && !platform.isIos()) {
-    return true;
+  // We do not support autoplay in amp-lite viewer regardless of platform.
+  if (isLiteViewer) {
+    return supportsAutoplayCache_ = Promise.resolve(false);
   }
 
+  // Short-circuit for known unsupported versions on mobile.
   const version = platform.getMajorVersion();
-  if (platform.isAndroid()) {
-    if (platform.isFirefox() && version >= 37) {
-      return true;
-    }
-    if (platform.isChrome() && version >= 53) {
-      return true;
-    }
+  if (platform.isAndroid() && platform.isChrome() && version < 53) {
+    return supportsAutoplayCache_ = Promise.resolve(false);
   }
 
-  if (platform.isIos()) {
-    if (platform.isSafari() && version >= 10) {
-      return true;
-    }
+  if (platform.isIos() && platform.isSafari() && version < 10) {
+    return supportsAutoplayCache_ = Promise.resolve(false);
   }
 
-  // TODO(aghassemi): Test other combinations and add support.
-  // TODO(aghassemi): Is there a way to detect that autoplay has been disabled
-  // in the user preferences of the browser?
-  return false;
+  // To detect autoplay, we create a video element with a inline data URI
+  // and call `play()` on it.
+  const detectionElement = ampdoc.win.document.createElement('video');
+  detectionElement.setAttribute('muted', '');
+  detectionElement.setAttribute('playsinline', '');
+  detectionElement.setAttribute('webkit-playsinline', '');
+  detectionElement.setAttribute('height', '0');
+  detectionElement.setAttribute('width', '0');
+  setStyles(detectionElement, {
+    position: 'fixed',
+    top: '0',
+    width: '0',
+    height: '0',
+    opacity: '0',
+  });
+
+  // The following video file is an h.264 encoded two-second video with four
+  // black frames generated with wide decoding compatibility in mind.
+  // File is 909 bytes (1212 in base64).
+  // Video file is generated by the following `ffmpeg` command:
+  //  $ffmpeg -f lavfi -i color=color=black:rate=2:size=100x100 -t 2
```

Comment:
Difference is only 74 bytes. 1138 instead of 1212 which makes sense since despite video being 2 seconds, it is only 2 frames per second so changing duration won't do much. Anyway I really don't want to rely on an `edge` case for this test video. (Also note that we are moving this out of `v0.js`)